source/material_model/compositing.cc

```diff
@@ -0,0 +1,207 @@
+        {
+          models[i]->evaluate(in, base_output);
+          copy_required_properties(i, base_output, out);
+        }
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::declare_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Compositing");
+        {
+          std::map<std::string, Property::MaterialProperty>::const_iterator prop_it = Property::property_map.begin();
+          for (; prop_it != Property::property_map.end(); ++prop_it)
+            {
+              prm.declare_entry(prop_it->first, "compositing",
+                                Patterns::Selection(MaterialModel::get_valid_model_names_pattern<dim>()),
+                                "Material model to use for " + prop_it->first +". Valid values for this "
+                                "parameter are the names of models that are also valid for the "
+                                "``Material models/Model name'' parameter. See the documentation for "
+                                "that for more information.");
+            }
+        }
+        prm.leave_subsection();
+      }
+      prm.leave_subsection();
+    }
+
+    template <int dim>
+    void
+    Compositing<dim>::parse_parameters (ParameterHandler &prm)
+    {
+      prm.enter_subsection("Material model");
+      {
+        prm.enter_subsection("Compositing");
+        {
+          model_names.clear();
+          std::map<std::string, Property::MaterialProperty>::const_iterator prop_it = Property::property_map.begin();
+          for (; prop_it != Property::property_map.end(); ++prop_it)
+            {
+              const Property::MaterialProperty prop = prop_it->second;
+              const std::string model_name = prm.get(prop_it->first);
+
+              AssertThrow(model_name != "averaging",
+                          ExcMessage("You may not use ``averaging'' as the base model for the "
+                                     + prop_it->first +" property of a compositing material model."));
+              AssertThrow(model_name != "compositing",
+                          ExcMessage("You may not use ``compositing'' as the base model for the "
+                                     + prop_it->first +" property of a compositing material model."));
+              unsigned int model_ind;
+              for (model_ind=0; model_ind < model_names.size(); ++model_ind)
+                {
+                  if (model_names[model_ind] == model_name)
+                    break;
+                }
+              if (model_ind == model_names.size())
+                model_names.push_back(model_name);
+
+              model_property_map[prop] = model_ind;
+            }
+          models.resize(model_names.size());
+
+          // create the model and initialize their SimulatorAccess base
+          // class; it will get a chance to read its parameters below after we
+          // leave the current section
```

Comment:
I see why you can not do the parsing here, but can you just move these lines below out of the subsection? It would be easier to understand if there were only one loop over the model names.